### Outliers

In [1]:
import Utility

In [2]:
Utility.path = './corpus/a/'

In [3]:
import pandas as pd

limit = 4 # 4 most frequent words (MFWs)
samples = ['Gratian1', 'dePen', 'Gratian2']
unknown = 'Gratian0'
features = Utility.get_features(samples)
# mfws = list(features.keys())[:limit]
mfws = ['in', 'non', 'an', 'si', 'quomodo']
counts = Utility.get_counts(mfws, [unknown] + samples)
lengths = Utility.get_lengths([unknown] + samples)
frequencies = (counts / lengths.values) * 1000
means = frequencies[samples].mean(axis = 1).to_frame('mean')
standard_deviations = frequencies[samples].std(axis = 1).to_frame('std')
z_scores = (frequencies - means.values) / standard_deviations.values
pd.options.display.float_format = '{:,.4f}'.format
means

mean
in      26.4656
non     24.0765
an       1.3815
si       9.2665
quomodo  1.0809

In [4]:
standard_deviations

std
in      2.0691
non     2.6598
an      0.5011
si      2.6245
quomodo 1.4119

In [5]:
frequencies

Gratian0  Gratian1   dePen  Gratian2
in        20.5270   25.5673 24.9975   28.8320
non        6.6574   23.9804 26.7831   21.4662
an        39.1123    1.3577  0.8928    1.8941
si        14.4244    6.2420 10.6140   10.9435
quomodo    0.8322    0.5642  2.6783    0.0000

In [6]:
z_scores

Gratian0  Gratian1   dePen  Gratian2
in        -2.8702   -0.4342 -0.7095    1.1437
non       -6.5491   -0.0361  1.0176   -0.9814
an        75.2996   -0.0475 -0.9754    1.0229
si         1.9653   -1.1524  0.5134    0.6390
quomodo   -0.1761   -0.3659  1.1314   -0.7655

The frequency of occurrence of the word *an* in the Gratian0 sample representing the thirty-six hypothetical case statements (*themata*) is a remarkably high 39.1123 occurrences per 1,000 words. By way of comparison, the mean frequency of occurrence of *an* across the three samples representing the first-recension *dicta* excluding *de Penitentia* (Gratian1), the first- and second-recension *dicta* from *de Penitentia* (dePen), and the second recension *dicta* (Gratian2) is 1.3815 occurrences per 1,000 words with a sample standard deviation of 0.5011. The frequency of occurrence of *an* in the Gratian0 sample is therefore 75.2996 standard deviations away from the mean frequency of occurrence of the same word in the Gratian1, dePen, and Gratian2 samples. The frequency of occurrence of the word *si* in the Gratian0 sample, on the other hand, is 14.4244 occurrences per 1,000 words. Given that the mean frequency of occurrence of *si* across the Gratian1, dePen, and Gratian2 samples is 9.2665 occurrences per 1,000 words and that the sample standard deviation is 2.6245, the frequency of occurrence of *si* in the Gratian0 sample is far less of an outlier at 1.9653 standard deviations away from the mean than *an* was.